### PR TITLE
add subregion (admin2Codes)

### PR DESCRIPTION
**Pull request by Guts**
See #159

### CODE REVIEW

File: cities_light/abstract_models.py [hunk 1]
```diff
@@ -160,6 +185,10 @@ class AbstractCity(Base):
         null=True,
         blank=True)
 
+    # subregion = models.CharField(max_length=10, null=True, blank=True,
+    #                              db_index=True)
```
Comment (jpic) [inline cities_light/abstract_models.py]: Can we remove commented code ? ;)

File: cities_light/management/commands/cities_light.py [hunk 2]
```diff
@@ -260,6 +266,20 @@ def _get_region_id(self, country_code2, region_id):
...
+        Simple lazy identity map for (country_code2, region_id, subregion_id)->subregion
+        """
+        country_id = self._get_country_id(country_code2)
+        if region_id not in self._region_codes[country_id]:
+            self._region_codes[country_id][region_id] = Region.objects.get(
+                country_id=country_id, geoname_code=region_id).pk
+        if subregion_id not in self._subregion_codes[country_id]:
+            self._subregion_codes[country_id][subregion_id] = SubRegion.objects.get(
+                country_id=country_id, geoname_code=subregion_id).pk
+        print(self._subregion_codes)
```
Comment (jpic) [inline cities_light/management/commands/cities_light.py]: Can we remove this print statement ?

File: cities_light/management/commands/cities_light.py [hunk 3]
```diff
@@ -360,6 +380,79 @@ def region_import(self, items):
...
+        try:
+            force_insert = False
+            force_update = False
+            subregion = SubRegion.objects.get(geoname_id=items[ISubRegion.geonameid])
+            force_update = True
+        except SubRegion.DoesNotExist:
+            if self.noinsert:
+                return
+            subregion = SubRegion(geoname_id=items[ISubRegion.geonameid])
+            force_insert = True
```
Comment (jpic) [inline cities_light/management/commands/cities_light.py]: Have you considered something like this ?

```
force_insert = force_update = False
subregion = SubRegion.objects.filter(geoname_id=items[ISubRegion.geonameid]).first()
if subregion:
    force_update = True
elif not subregion and self.noinsert:
    return
else:
    subregion = SubRegion(geoname_id=items[ISubRegion.geonameid])
    force_insert = True
```

File: cities_light/migrations/0001_squashed_0013_auto_20171229_1942.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        migrations.AddField(
+            model_name='city',
+            name='subregion',
+            field=models.ForeignKey(blank=True, null=True, on_delete=django.db.models.deletion.CASCADE, to='cities_light.SubRegion'),
+        ),
+        migrations.AlterUniqueTogether(
+            name='subregion',
+            unique_together=set([('country', 'slug'), ('country', 'name')]),
+        ),
+    ]
```
Comment (jpic) [inline cities_light/migrations/0001_squashed_0013_auto_20171229_1942.py]: Can we squash migrations in another, prior commit ?

File: cities_light/migrations/0009_auto_20171229_1656.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('cities_light', '0008_city_timezone'),
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name='Department',
```
Comment (jpic) [inline cities_light/migrations/0009_auto_20171229_1656.py]: I like that it's the same word as in my language but should this not be SubRegion ?

File: cities_light/migrations/0013_auto_20171229_1942.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    dependencies = [
+        ('cities_light', '0012_auto_20171229_1839'),
+    ]
+
+    operations = [
+        migrations.AlterUniqueTogether(
+            name='subregion',
+            unique_together=set([('country', 'name'), ('country', 'slug')]),
+        ),
+    ]
```
Comment (jpic) [inline cities_light/migrations/0013_auto_20171229_1942.py]: Could you try to redo migrations perhaps with something as such:

rm -rf cities_light/migrations
git reset --soft origin/master
git checkout cities_light/migrations
manage.py makemigrations cities_light -n add_subregion
git add cities_light/migrations
git commit -m "Add SubRegion feature"
git push -f

File: cities_light/settings.py [hunk 7]
```diff
@@ -99,24 +99,26 @@
...
     ['http://download.geonames.org/export/dump/admin1CodesASCII.txt'])
+SUBREGION_SOURCES = getattr(settings, 'CITIES_LIGHT_SUBREGION_SOURCES',
+    ['http://download.geonames.org/export/dump/admin2Codes.txt'])
 CITY_SOURCES = getattr(settings, 'CITIES_LIGHT_CITY_SOURCES',
     ['http://download.geonames.org/export/dump/cities15000.zip'])
 TRANSLATION_SOURCES = getattr(settings, 'CITIES_LIGHT_TRANSLATION_SOURCES',
     ['http://download.geonames.org/export/dump/alternateNames.zip'])
 TRANSLATION_LANGUAGES = getattr(settings, 'CITIES_LIGHT_TRANSLATION_LANGUAGES',
-    ['es', 'en', 'pt', 'de', 'pl', 'abbr'])
+    ['es', 'en', 'fr', 'pt', 'de', 'pl', 'abbr'])
```
Comment (jpic) [inline cities_light/settings.py]: :joy:  :joy:  :joy:

File: cities_light/settings.py [hunk 8]
```diff
@@ -139,12 +141,12 @@
 if INDEX_SEARCH_NAMES is None:
     INDEX_SEARCH_NAMES = True
     for database in list(settings.DATABASES.values()):
-        if 'mysql' in database['ENGINE'].lower():
+        if "ENGINE" in database and 'mysql' in database.get('ENGINE').lower():
```
Comment (jpic) [inline cities_light/settings.py]: Nice, you got the #167 #168 here ;)

File: cities_light/abstract_models.py (hunk 1 above)
Comment (Guts) [reply cities_light/abstract_models.py]: of course!

File: cities_light/settings.py (hunk 7 above)
Comment (Guts) [reply cities_light/settings.py]: Ups, I'v forgotten to undo this change!

File: cities_light/migrations/0013_auto_20171229_1942.py (hunk 6 above)
Comment (Guts) [reply cities_light/migrations/0013_auto_20171229_1942.py]: Ok I tried something but I'm not really a git pro so I hope that's not worse now than before...

File: cities_light/management/commands/cities_light.py (hunk 3 above)
Comment (Guts) [reply cities_light/management/commands/cities_light.py]: Not really. I let you improve or fix things as you think it's better.

**Comment by Guts** [comment]
@jpic I tried to fix some of improvements you suggested but I'm not sure that I can finish everything by myself.
Thanks for your positive comments and happy new year :)

**Comment by Guts** [comment]
I messed up the git branch. I give up on this. I'm going to delete my branch, this PR and i'll be back with a fresh one!

**Comment by jpic** [comment]
Awesome ! Can't wait ;)

**Comment by jpic** [comment]
Don't forget to reping as I forgot to test it me thinks ^^

**Pull request closed without merging**